apollo-router-core/src/plugins/csrf.rs

```diff
@@ -0,0 +1,335 @@
+///
+/// We don't want random websites to be able to execute actual GraphQL operations
+/// from a user's browser unless our CORS policy supports it. It's not good
+/// enough just to ensure that the browser can't read the response from the
+/// operation; we also want to prevent CSRF, where the attacker can cause side
+/// effects with an operation or can measure the timing of a read operation. Our
+/// goal is to ensure that we don't run the context function or execute the
+/// GraphQL operation until the browser has evaluated the CORS policy, which
+/// means we want all operations to be pre-flighted. We can do that by only
+/// processing operations that have at least one header set that appears to be
+/// manually set by the JS code rather than by the browser automatically.
+///
+/// POST requests generally have a content-type `application/json`, which is
+/// sufficient to trigger preflighting. So we take extra care with requests that
+/// specify no content-type or that specify one of the three non-preflighted
+/// content types. For those operations, we require (if this feature is enabled)
+/// one of a set of specific headers to be set. By ensuring that every operation
+/// either has a custom content-type or sets one of these headers, we know we
+/// won't execute operations at the request of origins who our CORS policy will
+/// block.
+#[derive(Debug, Clone)]
+struct Csrf {
+    config: CSRFConfig,
+}
+
+#[async_trait::async_trait]
+impl Plugin for Csrf {
+    type Config = CSRFConfig;
+
+    async fn new(config: Self::Config) -> Result<Self, BoxError> {
+        Ok(Csrf { config })
+    }
+
+    fn router_service(
+        &mut self,
+        service: BoxService<RouterRequest, RouterResponse, BoxError>,
+    ) -> BoxService<RouterRequest, RouterResponse, BoxError> {
+        if !self.config.unsafe_disabled {
+            let required_headers = self.config.required_headers.clone();
+            ServiceBuilder::new()
+                .checkpoint(move |req: RouterRequest| {
+                    if is_preflighted(&req, required_headers.as_slice()) {
+                        Ok(ControlFlow::Continue(req))
+                    } else {
+                        let error = crate::Error {
+                            message: format!("This operation has been blocked as a potential Cross-Site Request Forgery (CSRF). \
+                            Please either specify a 'content-type' header (with a mime-type that is not one of {}) \
+                            or provide one of the following headers: {}", 
+                            NON_PREFLIGHTED_CONTENT_TYPES.join(", "),
+                            required_headers.join(", ")),
+                            locations: Default::default(),
+                            path: Default::default(),
+                            extensions: Default::default(),
+                        };
```

Comment:
NIT: If think we have builders for errors

Reply:
nice catch! fixed in a20105b4